docs/design/scheduler.md

```diff
@@ -0,0 +1,46 @@
+
+# The Kubernetes Scheduler
+
+The Kubernetes scheduler runs as a process alongside the other master
+components such as the API server. Its interface to the API server is to watch
+for Pods with an empty PodSpec.NodeName, and for each them to post a Binding
+indicating where the Pod should be scheduled.
+
+## The scheduling process
+
+The scheduler tries to find a node for each Pod, one at a time, as it notices
+these Pods via watch. There are three steps. First it applies a set of "predicates" that filter out
+inappropriate nodes. For example, if the PodSpec specifies resource limits, then the scheduler
+will filter out nodes that don't have at least that much resources available (computed
+as the capacity of the node minus the sum of the resource limits of the Pods that
+are already running on the node). Second, it applies a set of "priority functions"
+that rank the nodes that weren't filtered out by the predicate check. For example,
+if the PodSpec specified resource limits, then it favors the least-loaded
+machine (measured in terms of the sum of the limits of the Pods already on each node).
+Finally, the node with the highest priority is chosen
+(or, if there are multiple such nodes, then one of them is chosen at random). The code
+for this main scheduling loop is in the function `Schedule()` in
+[plugin/pkg/scheduler/generic_scheduler.go](../../plugin/pkg/scheduler/generic_scheduler.go)
+
+## Scheduler extensibility
+
+The scheduler is extensible: the cluster administrator can choose which of the pre-defined
+scheduling policies to apply, and can add new ones. The built-in predicates and priorities are
+defined in [plugin/pkg/scheduler/predicates.go](../../plugin/pkg/scheduler/predicates.go) and
+[plugin/pkg/scheduler/priorities.go](../../plugin/pkg/scheduler/priorities.go), respectively.
+The policies that are applied when scheduling can be chosen in one of two ways. Normally,
+the policies used are selected by the functions `defaultPredicates()` and `defaultPriorities()` in
+[plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go](../../plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go).
+However, the choice of policies
+can be overridden by passing the command-line flag `--policy-config-file` to the scheduler, pointing to a JSON
+file specifying which scheduling policies to use. See
+[examples/scheduler-policy-config.json](../../examples/scheduler-policy-config.json) for an example
+config file. (Note that the config file format is versioned; the API is defined in
+[plugin/pkg/scheduler/api/](../../plugin/pkg/scheduler/api/)).
+Thus to add a new scheduling policy, you should modify predicates.go or priorities.go,
+and either register the policy in `defaultPredicates()` or `defaultPriorities()`, or use a policy config file.
+
+## Exploring the code
+
+If you want to get a global picture of how the scheduler work, you can start in
```

Comment:
"If you want to get a global picture of how the scheduler works, ......"

Hi, I think this is a great design doc for scheduler. I have two other comments above, but I think leaving the doc the way it is now is also a good idea.